src/strings.h

```diff
@@ -0,0 +1,184 @@
+    {
+        memset(s, 0, n);
+    }
+
+    static void explicit_bzero(void * s, size_t n)
+    {
+        volatile char * vs = (volatile char *) s;
+        while (n)
+        {
+            *vs++ = 0;
+            n--;
+        }
+    }
+
+    static char * index(const char * s, int c)
+    {
+        return (char*)strchr(s, c);
+    }
+
+    static char * rindex(const char * s, int c)
+    {
+        return (char*)strrchr(s, c);
+    }
+
+    static int ffs(int i)
+    {
+        int bit;
+
+        if (0 == i)
+            return 0;
+
+        for (bit = 1; !(i & 1); ++bit)
+            i >>= 1;
+        return bit;
+    }
+
+    static int ffsl(long i)
+    {
+        int bit;
+
+        if (0 == i)
+            return 0;
+
+        for (bit = 1; !(i & 1); ++bit)
+            i >>= 1;
+        return bit;
+    }
+
+    static int ffsll(long long i)
+    {
+        int bit;
+
+        if (0 == i)
+            return 0;
+
+        for (bit = 1; !(i & 1); ++bit)
+            i >>= 1;
+        return bit;
+    }
+
+#ifndef __MINGW32__
+
+    static int strcasecmp(const char * s1, const char * s2)
+    {
+        const unsigned char * u1 = (const unsigned char*) s1;
+        const unsigned char * u2 = (const unsigned char*) s2;
+        int result;
+
+        while ((result = tolower(*u1) - tolower(*u2)) == 0 && *u1 != 0)
+        {
+            *u1++;
+            *u2++;
+        }
+
+        return result;
+    }
+
+    static int strncasecmp(const char * s1, const char * s2, size_t	 n)
+    {
+        const unsigned char * u1 = (const unsigned char*) s1;
+        const unsigned char * u2 = (const unsigned char*) s2;
+        int result;
+
+        for (; n != 0; n--)
+        {
+            result = tolower(*u1) - tolower(*u2);
+            if (result)
+                return result;
+            if (*u1 == 0)
+                return 0;
+        }
+        return 0;
+    }
+
+    static int strcasecmp_l(const char * s1, const char * s2, _locale_t loc)
+    {
+        const unsigned char * u1 = (const unsigned char*) s1;
+        const unsigned char * u2 = (const unsigned char*) s2;
+        int result;
+
+        while ((result = _tolower_l(*u1, loc) - _tolower_l(*u2, loc)) == 0 && *u1 != 0)
+        {
+            *u1++;
+            *u2++;
+        }
+
+        return result;
+    }
+
+    static int strncasecmp_l(const char * s1, const char * s2, size_t n, _locale_t loc)
+    {
+        const unsigned char * u1 = (const unsigned char*) s1;
+        const unsigned char * u2 = (const unsigned char*) s2;
+        int result;
+
+        for (; n != 0; n--)
+        {
+            result = _tolower_l(*u1, loc) - _tolower_l(*u2, loc);
+            if (result)
+                return result;
+            if (*u1 == 0)
+                return 0;
+        }
+        return 0;
+    }
+
+
+#endif
+
+#ifdef __cplusplus
+}
+#endif /* __cplusplus */
+
+#endif /* _WIN32 */
+
+#endif /* __STRINGS_H_868520A6_2F4E_4B49_AF4B_02B5A21CF6BB__ */
```

Comment:
Isn't that provided by Microsoft C runtime? If not, it shall be moved into `contrib/`.

Reply:
Haha, oops, it is, I'll fix and remove that.